core/celo_multi_gaspool.go

```diff
@@ -0,0 +1,79 @@
+package core
+
+import "github.com/ethereum/go-ethereum/common"
+
+type FeeCurrency = common.Address
+
+// MultiGasPool tracks the amount of gas available during execution
+// of the transactions in a block per fee currency. The zero value is a pool
+// with zero gas available.
+type MultiGasPool struct {
+	pools       map[FeeCurrency]*GasPool
+	defaultPool *GasPool
+
+	blockGasLimit uint64
+	defaultLimit  float64
+}
+
+type FeeCurrencyLimitMapping = map[FeeCurrency]float64
+
+// NewMultiGasPool creates a multi-fee currency gas pool and a default fallback
+// pool for CELO
+func NewMultiGasPool(
+	blockGasLimit uint64,
+	defaultLimit float64,
+	limitsMapping FeeCurrencyLimitMapping,
+) *MultiGasPool {
+	pools := make(map[FeeCurrency]*GasPool, len(limitsMapping))
+	// A special case for CELO which doesn't have a limit
+	celoPool := new(GasPool).AddGas(blockGasLimit)
+	mgp := &MultiGasPool{
+		pools:         pools,
+		defaultPool:   celoPool,
+		blockGasLimit: blockGasLimit,
+		defaultLimit:  defaultLimit,
+	}
+	for feeCurrency, fraction := range limitsMapping {
+		mgp.getOrInitPool(feeCurrency, &fraction)
+	}
+	return mgp
+}
+
+func (mgp MultiGasPool) getOrInitPool(c FeeCurrency, fraction *float64) *GasPool {
+	if gp, ok := mgp.pools[c]; ok {
```

Comment:
#116 mentions the concurrent existence of FeeCurrencyAdapter and the adapted currency itself.

Do we need to consider this here as a n-to-1 mapping of addresses to gas-pools?
Because as it currently is implemented this means that both versions of the  fee currency have separate gas-pools and limits.

Reply:
That's great that you thought about this use case but I don't think we need to be that precise here and implement n-to-1 mapping, there is only a short period of time where both addresses would co-exist and can be handled by separate settings if needed.